content/en/docs/quarks-operator/concepts/ha.md

```diff
@@ -0,0 +1,182 @@
+
+In short, this uses labeled nodes and spawns `replica count` pods on each node.
+
+## Multi AZ
+
+In order for things to work correctly across versions and AZs, we need [ClusterIP `Services`](https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#using-stable-network-identities) that select for Instance Group `Pods`.
+
+Services:
+```text
+nats-0
+  selects pod z0-v1-0
+  selects pod z0-v2-0
+nats-1
+  selects pod z1-v1-0
+  selects pod z1-v2-0
+nats-2
+  selects pod z0-v1-1
+  selects pod z0-v2-1
+nats-3
+  selects pod z1-v1-1
+  selects pod z1-v2-1
+nats-4
+  selects pod z0-v1-2
+  selects pod z0-v2-2
+nats-5
+  selects pod z1-v1-2
+  selects pod z1-v2-2
+```
+(from [Services And DNS Addresses](rendering_templates.md#services-and-dns-addresses) )
+
+The services select pods via the pod-ordinal label.
+
+Service selector example for "nats-0" service:
+```yaml
+selector:
+  quarks.cloudfoundry.org/az-index: "0"
+  quarks.cloudfoundry.org/deployment-name: kubecf
+  quarks.cloudfoundry.org/instance-group-name: nats
+  quarks.cloudfoundry.org/pod-ordinal: "0"
+```
+
+Quarks also generates an additional `nats' service.
+
+
+# Ordinals
+
+Each pod in an instance group's statefulset has several "ordinals" set as labels.
+
+* pod-ordinal
+* startup-ordinal
+* spec-index
+* az-index
+
+## Pod Ordinal
+
+Quarks stores the pod ordinal, which is the [suffix from the pod name](https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#using-stable-network-identities), in a label.
+This is done by the pod mutator in Quarks statefulset.
+
+## Startup Ordinal
+
+Quarks also stores a startup ordinal in another label. Its value is 0 for the first pod to start in a statefulset revision. It is also persisted, so it stays the same if pods within a revision restart.
+
+That label is then referenced by the POD_ORDINAL env var in instance group pods. (There is special treatment for errands.)
+
+The POD_ORDINAL is used by the template-render sub command in an init container, to render the final version of the job configs from their eruby templates.
+
+## Spec Index
+
+Appears to be unused and is always equal to the pod ordinal.
+
+## AZ Index
+
+Used to calculate names and indexes for [multi AZ](from_bosh_to_kube.md#dealing-with-azs).
+
+It is calculated using the following formula:
+
+```text
+(AZ_INDEX - 1) * REPLICAS + POD_ORDINAL
+```
+
+# BOSH Job Spec
+
+Several BOSH mechanisms rely on the ordinals, which are derived from the statefulset state.
+
+I.e. the [job spec](https://bosh.io/docs/jobs/) is calculated with those numbers.
+
+Currently:
+
+* spec.bootstrap uses the startup-ordinal
+* spec.index uses the startup-ordinal
+* spec.address uses the startup-ordinal
+
+## Spec Bootstrap
+
+> True if this instance is the first instance of its group.
+([BOSH jobs](https://bosh.io/docs/jobs/))
+
+The pod that starts first in a statefulet must have the bootstrap flag. It's used to initialize databases and such.
```

Comment:
Given that each AZ has its own stateful set, and each stateful set manages its own startup ordinals, how does quarks determine which pod is the bootstrap pod? Will it always be from az-index 0?

Reply:
Yes, that's what I'm trying to figure out, so I can use startup-ordinal only for that...

Any pointers on how bootstrap is supposed to behave with multiple AZs? I guess, only run once per job, on the first pod in that is started. On update, in a new controller revision, run again on the first pod.

Reply:
> Any pointers on how bootstrap is supposed to behave with multiple AZs?

From reading various tracker stories, it looks like bosh would pick the first instance from the first az and mark it as the bootstrap instance. If that az was later removed, the bootstrap instance would move to the first still available zone.

This would be a "sticky" attribute of the instance; adding the original zone back to the cluster would *not* move the bootstrap instance back; the only time the bootstrap instance moves is when the current one is no longer available. See also https://www.pivotaltracker.com/n/projects/956238/stories/100245974

I do not understand why they consider it important that the bootstrap instance remains on the same VM, if at all possible. Maybe it is to avoid a race condition where you could temporarily have multiple bootstrap instances?

I think the mechanism documented so far (use first `instance.index` on initial deploy, and highest `instance.index` during update) should work fine for us.